### ISSUE TITLE
`GeneratorCall`s and other non-`Module` instantiables as parameters

### ISSUE THREAD

**Issue by dan-fritchman** [dan-fritchman/Hdl21#93]
This: 

```python
import hdl21 as h
from hdl21.generators import Pmos as Hdl21Pmos


@h.generator
def Pmos(_: h.HasNoParams) -> h.Module:
    return Hdl21Pmos()


import pickle
pickle.dumps(Pmos())
```

Fails with errors like so: 

```python
Traceback (most recent call last):
  File "huh.py", line 9, in <module>
    pickle.dumps(Pmos())
_pickle.PicklingError: Can't pickle <function Pmos at 0x7fa476a13d90>: it's not the same object as __main__.Pmos
```

Not the end of the world in a vacuum; we don't inherently need to be able to pickle-up every possible Hdl21 program.  
Problem is, we also use pickle internally, when comparing parameters, particularly as part of `Generator` equality-testing: 

```python
## from hdl21/ generator.py 

class GeneratorCall: 
    def __hash__(self):
        return hash((id(self.gen), pickle.dumps(self.params))) ## <= ruh roh 
```

Which makes it impossible in that prior snippet to do something like: 

```python
@h.paramclass
class Params:
  unit = h.Param(dtype=h.Instantiable, desc="the unit pmos")

@h.generator
def PmosPair(params: Params) -> h.Module:
  ... # you get it, stick two of `params.unit` into a diff pair
```

I believe the "pickle for equality" test was an @aviralpandey addition, and I believe for something very much like this use-case - using a `Module` as a parameter. (That sound right?) 

Assuming so, and there's not a bunch of other complications - probably recommend we make `Module` (and maybe a few similar types) "identity is equality" based instead. I.e. add this stuff, currently used on `Signal` and many others: 

```python 
class Module: ## and maybe a few others

    def __eq__(self, other) -> bool:
        # Identity is equality
        return id(self) == id(other)

    def __hash__(self) -> bool:
        # Identity is equality
        return hash(id(self))
```

**Comment by dan-fritchman** [comment]
FWIW we might rename this issue something like "`Module` works as a parameter, but a generated module (`GeneratorCall`) does not". 

Commit [32e05ba](https://github.com/dan-fritchman/Hdl21/pull/92/commits/32e05ba9da2b0c98bee3562d4960bf3f6020bc4c) does: 

* Adds the "identity is equality" comparison & hashing to `Module`, `Generator`, `ExternalModule`, and `Primitive`
* It also adds the "identity of functions, equality of parameters" comparisons to `ExternalModuleCall` and `PrimitiveCall`, making them like `GeneratorCall` in this respect. 

I think these are probably good ideas regardless of how we decide to resolve the pickling/ generator-parameter-serializing problems.  
That commit passes all our existing tests, but will need some more work to support the example above.

**Comment by dan-fritchman** [comment]
73a77f9 has more or less the snippet above as a unit test, which passes. 

I'm not sure what the "more work" I wrote about above was.  
Maybe other recent changes have made this work.  

That'll get merged in via #195.
